Add a ZT fastpath for linalg_cross

nkaretnikov opened the pull request:
See #69687.

upto 89000x speed up. see https://github.com/pytorch/pytorch/pull/76940#issuecomment-1120522858

Commit:
Add a ZT fastpath for linalg_cross

Also removes a redundant internal assert.

nkaretnikov commented:
test/benchmark:

```py
import torch

shapes = (
    (1, 4),  # broadcasting
    (4, 3),
    (4, 1),  # broadcasting
)

def tensors(s1, s2):
    return (
        # label, a, b
        (f"randn, randn", torch.randn(s1), torch.randn(s2)),
        (f"randn, zeros", torch.randn(s1), torch.zeros(s2)),
        (f"zeros, randn", torch.zeros(s1), torch.randn(s2)),
        (f"zeros, zeros", torch.zeros(s1), torch.zeros(s2)),
        (f"randn, zt   ", torch.randn(s1), torch._efficientzerotensor(s2)),
        (f"zt,    randn", torch._efficientzerotensor(s1), torch.randn(s2)),
    )

for s1 in shapes:
    for s2 in shapes:
        errors = set()

        print(s1, s2)

        for label, t1, t2 in tensors(s1, s2):
            print(label)

            with torch.profiler.profile(
                activities=[
                    torch.profiler.ProfilerActivity.CPU
                ]
            ) as p:
                try:
                    res = torch.linalg.cross(t1, t2)

                    print(res)  # result
                    print(res._is_zerotensor())  # is zt

                except RuntimeError as e:
                    # print(e)  # error
                    errors.add(str(e))

            # s = str(p.key_averages())  # stack trace
            s = str(p.key_averages()).split('\n')[-2]  # time
            print(s)

        assert len(errors) <= 1  # same error
        # print the error for this group of shapes
        if len(errors) > 0:
            print("error:", list(errors)[0])
        print()

```

conclusion: no difference vs zeros on cpu (or even slower sometimes)

<details>
  <summary>see results</summary>

```
(1, 4) (1, 4)
randn, randn
Self CPU time total: 3.200ms
randn, zeros
Self CPU time total: 2.936ms
zeros, randn
Self CPU time total: 2.932ms
zeros, zeros
Self CPU time total: 2.940ms
randn, zt   
Self CPU time total: 4.092ms
zt,    randn
Self CPU time total: 3.983ms
error: dimension -1 does not have size 3

(1, 4) (4, 3)
randn, randn
Self CPU time total: 3.211ms
randn, zeros
Self CPU time total: 3.196ms
zeros, randn
Self CPU time total: 3.207ms
zeros, zeros
Self CPU time total: 3.210ms
randn, zt   
Self CPU time total: 4.242ms
zt,    randn
Self CPU time total: 4.263ms
error: The size of tensor a (4) must match the size of tensor b (3) at non-singleton dimension 1

(1, 4) (4, 1)
randn, randn
Self CPU time total: 2.939ms
randn, zeros
Self CPU time total: 2.952ms
zeros, randn
Self CPU time total: 2.936ms
zeros, zeros
Self CPU time total: 2.949ms
randn, zt   
Self CPU time total: 3.989ms
zt,    randn
Self CPU time total: 3.975ms
error: dimension -1 does not have size 3

(4, 3) (1, 4)
randn, randn
Self CPU time total: 3.208ms
randn, zeros
Self CPU time total: 3.204ms
zeros, randn
Self CPU time total: 3.195ms
zeros, zeros
Self CPU time total: 3.207ms
randn, zt   
Self CPU time total: 4.253ms
zt,    randn
Self CPU time total: 4.240ms
error: The size of tensor a (3) must match the size of tensor b (4) at non-singleton dimension 1

(4, 3) (4, 3)
randn, randn
tensor([[-0.9556,  0.1290,  0.4093],
        [-1.5410,  2.4429,  3.5799],
        [ 0.3877,  0.5162,  0.2456],
        [-0.1289,  0.6153,  0.2188]])
False
Self CPU time total: 458.000us
randn, zeros
tensor([[0., 0., -0.],
        [0., 0., -0.],
        [0., -0., 0.],
        [-0., 0., 0.]])
False
Self CPU time total: 71.000us
zeros, randn
tensor([[-0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., -0., 0.]])
False
Self CPU time total: 66.000us
zeros, zeros
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
False
Self CPU time total: 69.000us
randn, zt   
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
True
Self CPU time total: 89.000us
zt,    randn
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
True
Self CPU time total: 79.000us

(4, 3) (4, 1)
randn, randn
tensor([[ 0.1791, -0.0091, -0.1700],
        [ 0.7592, -0.8719,  0.1127],
        [ 1.0137, -2.2103,  1.1967],
        [ 0.8497,  1.0640, -1.9137]])
False
Self CPU time total: 137.000us
randn, zeros
tensor([[0., 0., -0.],
        [0., 0., -0.],
        [0., 0., -0.],
        [0., 0., 0.]])
False
Self CPU time total: 67.000us
zeros, randn
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
False
Self CPU time total: 69.000us
zeros, zeros
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
False
Self CPU time total: 67.000us
randn, zt   
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
True
Self CPU time total: 80.000us
zt,    randn
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
True
Self CPU time total: 80.000us

(4, 1) (1, 4)
randn, randn
Self CPU time total: 2.976ms
randn, zeros
Self CPU time total: 2.955ms
zeros, randn
Self CPU time total: 2.947ms
zeros, zeros
Self CPU time total: 2.937ms
randn, zt   
Self CPU time total: 3.967ms
zt,    randn
Self CPU time total: 3.963ms
error: dimension -1 does not have size 3

(4, 1) (4, 3)
randn, randn
tensor([[-0.5289,  0.2484,  0.2805],
        [-0.0444, -0.0861,  0.1305],
        [-0.0626, -0.3902,  0.4528],
        [ 0.2039, -0.3164,  0.1125]])
False
Self CPU time total: 143.000us
randn, zeros
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
False
Self CPU time total: 66.000us
zeros, randn
tensor([[0., 0., 0.],
        [-0., 0., 0.],
        [-0., 0., 0.],
        [-0., 0., 0.]])
False
Self CPU time total: 66.000us
zeros, zeros
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
False
Self CPU time total: 65.000us
randn, zt   
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
True
Self CPU time total: 77.000us
zt,    randn
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
True
Self CPU time total: 79.000us

(4, 1) (4, 1)
randn, randn
Self CPU time total: 2.915ms
randn, zeros
Self CPU time total: 2.897ms
zeros, randn
Self CPU time total: 2.881ms
zeros, zeros
Self CPU time total: 2.961ms
randn, zt   
Self CPU time total: 3.884ms
zt,    randn
Self CPU time total: 3.893ms
error: dimension -1 does not have size 3
```
</details>

nkaretnikov commented:
note: added as a separate kernel because i'm not sure how to do it with the structured one. the structured one is defined via the out variant, which means i cannot write to out (it's const) unless i use some unsafe data_ptr hackery. or is there a proper way to do it in IMPL there before dispatching via a stub?

nkaretnikov commented:
in the benchmark, the shape sanity checks are just to make sure i go thru the same code path on error. and the results are printed to see that i actually return initialized tensors instead of random values

anjali411 commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -699,6 +699,30 @@ Tensor sub_zerotensor(const Tensor& self, const Tensor& other, const Scalar& alp
+    other.to(device),
+    dim);
+
+  if (input._is_zerotensor() || other._is_zerotensor()) {
```
but this condition will always be true, right? we only call `linalg_cross_zerotensor` when the `ZeroTensor` dispatch key is set on one of the inputs

anjali411 commented:
It would be interesting to benchmark on CUDA for a range of sizes (small and big tensors). It is totally possible that calling `.to(device='meta')` multiple times and dispatching to meta kernel is not cheap.

anjali411 commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -699,6 +699,30 @@ Tensor sub_zerotensor(const Tensor& self, const Tensor& other, const Scalar& alp
+  const Tensor& input,
+  const Tensor& other,
+  const int64_t dim)
+{
```
Add a check to ensure that after broadcasting, input.size(dim) != 3 or other.size(dim) != 3

nkaretnikov replied:
not totally sure, but seems to be already checked in `TORCH_PRECOMPUTE_META_FUNC(linalg_cross)`, no?

```cpp
  TORCH_CHECK(input_broadcasted.size(dim) == 3, "dimension ", dimension, " does not have size 3");
```

you can double-check with this:

```py
import torch

a = torch.randn(4, 3)
b = torch.randn(4, 3)
# b = torch._efficientzerotensor(4, 3)
res = torch.linalg.cross(a, b)
print(res)

# a = torch.randn(1, 3)  # a is broadcast to match shape of b
a = torch._efficientzerotensor(1, 3)
res = torch.linalg.cross(a, b, dim=0)
print(res)

```

output:

```
RuntimeError: dimension 0 does not have size 3
```

anjali411 replied:
oh right okay so this gets checked because of this call below:
```
auto meta_out = at::redispatch::linalg_cross(
    c10::DispatchKeySet(at::DispatchKey::Meta),
    input.to(device),
    other.to(device),
    dim);
```

ok SGTM

anjali411 commented:
> note: added as a separate kernel because i'm not sure how to do it with the structured one. the structured one is defined via the out variant, which means i cannot write to out (it's const) unless i use some unsafe data_ptr hackery. or is there a proper way to do it in IMPL there before dispatching via a stub?

registering a new kernel is the recommended way.

nkaretnikov commented:
updated benchmark to test on cpu/cuda.

it does seem to make a difference on cuda with bigger tensors: see the very last case in the output below.
but also the perf sometimes degrades on cpu with zt. i'm not sure why (the 197.000us case below).

```py
import torch

devices = (
    'cpu',
    'cuda',
)

shapes = (
    (1, 4),  # broadcasting
    (4, 3),
    (4, 1),  # broadcasting
    (2, 4, 8, 16, 32, 64, 3),
)

def tensors(s1, s2, device):
    return (
        # label, a, b
        (f"{device}, randn, randn", torch.randn(s1, device=device), torch.randn(s2, device=device)),
        (f"{device}, randn, zeros", torch.randn(s1, device=device), torch.zeros(s2, device=device)),
        (f"{device}, zeros, randn", torch.zeros(s1, device=device), torch.randn(s2, device=device)),
        (f"{device}, zeros, zeros", torch.zeros(s1, device=device), torch.zeros(s2, device=device)),
        (f"{device}, randn, zt   ", torch.randn(s1, device=device), torch._efficientzerotensor(s2, device=device)),
        (f"{device}, zt,    randn", torch._efficientzerotensor(s1, device=device), torch.randn(s2, device=device)),
    )

for device in devices:
    for s1 in shapes:
        for s2 in shapes:
            errors = set()

            print(s1, s2)

            for label, t1, t2 in tensors(s1, s2, device):
                print(label)

                if device == 'cpu':
                    activity =  torch.profiler.ProfilerActivity.CPU
                elif device == 'cuda':
                    activity = torch.profiler.ProfilerActivity.CUDA
                else:
                    raise Exception(f'unexpected device: {device}')

                with torch.profiler.profile(
                    activities=[
                        activity
                    ]
                ) as p:
                    try:
                        res = torch.linalg.cross(t1, t2)

                        # print(res)  # result
                        print(res._is_zerotensor())  # is zt

                    except RuntimeError as e:
                        # print(e)  # error
                        errors.add(str(e))

                # s = str(p.key_averages())  # stack trace
                s = str(p.key_averages()).split('\n')[-2]  # time
                print(s)

            assert len(errors) <= 1  # same error
            # print the error for this group of shapes
            if len(errors) > 0:
                print("error:", list(errors)[0])
            print()
```

```
(2, 4, 8, 16, 32, 64, 3) (2, 4, 8, 16, 32, 64, 3)
cpu, randn, randn
False
Self CPU time total: 26.025ms
cpu, randn, zeros
False
Self CPU time total: 24.682ms
cpu, zeros, randn
False
Self CPU time total: 24.252ms
cpu, zeros, zeros
False
Self CPU time total: 23.695ms
cpu, randn, zt   
True
Self CPU time total: 197.000us
cpu, zt,    randn
True
Self CPU time total: 25.000us

(2, 4, 8, 16, 32, 64, 3) (2, 4, 8, 16, 32, 64, 3)
cuda, randn, randn
False
Self CUDA time total: 252.000us
cuda, randn, zeros
False
Self CUDA time total: 251.000us
cuda, zeros, randn
False
Self CUDA time total: 251.000us
cuda, zeros, zeros
False
Self CUDA time total: 251.000us
cuda, randn, zt   
True
Self CUDA time total: 2.000us
cuda, zt,    randn
True
Self CUDA time total: 2.000us
```

<details>
  <summary>see all results</summary>

```
(1, 4) (1, 4)
cpu, randn, randn
Self CPU time total: 3.268ms
cpu, randn, zeros
Self CPU time total: 2.975ms
cpu, zeros, randn
Self CPU time total: 2.971ms
cpu, zeros, zeros
Self CPU time total: 2.974ms
cpu, randn, zt   
Self CPU time total: 4.155ms
cpu, zt,    randn
Self CPU time total: 4.049ms
error: dimension -1 does not have size 3

(1, 4) (4, 3)
cpu, randn, randn
Self CPU time total: 3.252ms
cpu, randn, zeros
Self CPU time total: 3.244ms
cpu, zeros, randn
Self CPU time total: 3.228ms
cpu, zeros, zeros
Self CPU time total: 3.238ms
cpu, randn, zt   
Self CPU time total: 4.275ms
cpu, zt,    randn
Self CPU time total: 4.296ms
error: The size of tensor a (4) must match the size of tensor b (3) at non-singleton dimension 1

(1, 4) (4, 1)
cpu, randn, randn
Self CPU time total: 2.955ms
cpu, randn, zeros
Self CPU time total: 3.061ms
cpu, zeros, randn
Self CPU time total: 2.955ms
cpu, zeros, zeros
Self CPU time total: 2.957ms
cpu, randn, zt   
Self CPU time total: 4.143ms
cpu, zt,    randn
Self CPU time total: 4.161ms
error: dimension -1 does not have size 3

(1, 4) (2, 4, 8, 16, 32, 64, 3)
cpu, randn, randn
Self CPU time total: 3.363ms
cpu, randn, zeros
Self CPU time total: 3.285ms
cpu, zeros, randn
Self CPU time total: 3.280ms
cpu, zeros, zeros
Self CPU time total: 3.279ms
cpu, randn, zt   
Self CPU time total: 4.380ms
cpu, zt,    randn
Self CPU time total: 4.354ms
error: The size of tensor a (4) must match the size of tensor b (3) at non-singleton dimension 6

(4, 3) (1, 4)
cpu, randn, randn
Self CPU time total: 3.351ms
cpu, randn, zeros
Self CPU time total: 3.292ms
cpu, zeros, randn
Self CPU time total: 3.379ms
cpu, zeros, zeros
Self CPU time total: 3.273ms
cpu, randn, zt   
Self CPU time total: 4.817ms
cpu, zt,    randn
Self CPU time total: 4.530ms
error: The size of tensor a (3) must match the size of tensor b (4) at non-singleton dimension 1

(4, 3) (4, 3)
cpu, randn, randn
False
Self CPU time total: 27.000us
cpu, randn, zeros
False
Self CPU time total: 11.000us
cpu, zeros, randn
False
Self CPU time total: 11.000us
cpu, zeros, zeros
False
Self CPU time total: 10.000us
cpu, randn, zt   
True
Self CPU time total: 32.000us
cpu, zt,    randn
True
Self CPU time total: 24.000us

(4, 3) (4, 1)
cpu, randn, randn
False
Self CPU time total: 11.000us
cpu, randn, zeros
False
Self CPU time total: 11.000us
cpu, zeros, randn
False
Self CPU time total: 11.000us
cpu, zeros, zeros
False
Self CPU time total: 11.000us
cpu, randn, zt   
True
Self CPU time total: 24.000us
cpu, zt,    randn
True
Self CPU time total: 23.000us

(4, 3) (2, 4, 8, 16, 32, 64, 3)
cpu, randn, randn
Self CPU time total: 3.374ms
cpu, randn, zeros
Self CPU time total: 3.423ms
cpu, zeros, randn
Self CPU time total: 3.286ms
cpu, zeros, zeros
Self CPU time total: 3.300ms
cpu, randn, zt   
Self CPU time total: 4.429ms
cpu, zt,    randn
Self CPU time total: 4.383ms
error: The size of tensor a (4) must match the size of tensor b (64) at non-singleton dimension 5

(4, 1) (1, 4)
cpu, randn, randn
Self CPU time total: 3.030ms
cpu, randn, zeros
Self CPU time total: 3.133ms
cpu, zeros, randn
Self CPU time total: 3.019ms
cpu, zeros, zeros
Self CPU time total: 3.064ms
cpu, randn, zt   
Self CPU time total: 4.259ms
cpu, zt,    randn
Self CPU time total: 4.127ms
error: dimension -1 does not have size 3

(4, 1) (4, 3)
cpu, randn, randn
False
Self CPU time total: 14.000us
cpu, randn, zeros
False
Self CPU time total: 11.000us
cpu, zeros, randn
False
Self CPU time total: 10.000us
cpu, zeros, zeros
False
Self CPU time total: 11.000us
cpu, randn, zt   
True
Self CPU time total: 27.000us
cpu, zt,    randn
True
Self CPU time total: 23.000us

(4, 1) (4, 1)
cpu, randn, randn
Self CPU time total: 3.050ms
cpu, randn, zeros
Self CPU time total: 3.146ms
cpu, zeros, randn
Self CPU time total: 3.131ms
cpu, zeros, zeros
Self CPU time total: 3.039ms
cpu, randn, zt   
Self CPU time total: 4.092ms
cpu, zt,    randn
Self CPU time total: 4.093ms
error: dimension -1 does not have size 3

(4, 1) (2, 4, 8, 16, 32, 64, 3)
cpu, randn, randn
Self CPU time total: 3.393ms
cpu, randn, zeros
Self CPU time total: 3.295ms
cpu, zeros, randn
Self CPU time total: 3.309ms
cpu, zeros, zeros
Self CPU time total: 3.293ms
cpu, randn, zt   
Self CPU time total: 4.418ms
cpu, zt,    randn
Self CPU time total: 4.388ms
error: The size of tensor a (4) must match the size of tensor b (64) at non-singleton dimension 5

(2, 4, 8, 16, 32, 64, 3) (1, 4)
cpu, randn, randn
Self CPU time total: 3.386ms
cpu, randn, zeros
Self CPU time total: 3.326ms
cpu, zeros, randn
Self CPU time total: 3.295ms
cpu, zeros, zeros
Self CPU time total: 3.286ms
cpu, randn, zt   
Self CPU time total: 4.414ms
cpu, zt,    randn
Self CPU time total: 4.384ms
error: The size of tensor a (3) must match the size of tensor b (4) at non-singleton dimension 6

(2, 4, 8, 16, 32, 64, 3) (4, 3)
cpu, randn, randn
Self CPU time total: 3.414ms
cpu, randn, zeros
Self CPU time total: 3.317ms
cpu, zeros, randn
Self CPU time total: 3.298ms
cpu, zeros, zeros
Self CPU time total: 3.284ms
cpu, randn, zt   
Self CPU time total: 4.374ms
cpu, zt,    randn
Self CPU time total: 4.360ms
error: The size of tensor a (64) must match the size of tensor b (4) at non-singleton dimension 5

(2, 4, 8, 16, 32, 64, 3) (4, 1)
cpu, randn, randn
Self CPU time total: 3.404ms
cpu, randn, zeros
Self CPU time total: 3.303ms
cpu, zeros, randn
Self CPU time total: 3.294ms
cpu, zeros, zeros
Self CPU time total: 3.300ms
cpu, randn, zt   
Self CPU time total: 4.410ms
cpu, zt,    randn
Self CPU time total: 4.424ms
error: The size of tensor a (64) must match the size of tensor b (4) at non-singleton dimension 5

(2, 4, 8, 16, 32, 64, 3) (2, 4, 8, 16, 32, 64, 3)
cpu, randn, randn
False
Self CPU time total: 22.422ms
cpu, randn, zeros
False
Self CPU time total: 22.135ms
cpu, zeros, randn
False
Self CPU time total: 10.730ms
cpu, zeros, zeros
False
Self CPU time total: 10.069ms
cpu, randn, zt   
True
Self CPU time total: 68.000us
cpu, zt,    randn
True
Self CPU time total: 26.000us

(1, 4) (1, 4)
cuda, randn, randn
Self CPU time total: 9.000us
cuda, randn, zeros
Self CPU time total: 3.000us
cuda, zeros, randn
Self CPU time total: 3.000us
cuda, zeros, zeros
Self CPU time total: 3.000us
cuda, randn, zt   
Self CUDA time total: 2.000us
cuda, zt,    randn
Self CUDA time total: 2.000us
error: dimension -1 does not have size 3

(1, 4) (4, 3)
cuda, randn, randn
Self CPU time total: 4.000us
cuda, randn, zeros
Self CPU time total: 3.000us
cuda, zeros, randn
Self CPU time total: 3.000us
cuda, zeros, zeros
Self CPU time total: 3.000us
cuda, randn, zt   
Self CUDA time total: 2.000us
cuda, zt,    randn
Self CUDA time total: 2.000us
error: The size of tensor a (4) must match the size of tensor b (3) at non-singleton dimension 1

(1, 4) (4, 1)
cuda, randn, randn
Self CPU time total: 3.000us
cuda, randn, zeros
Self CPU time total: 3.000us
cuda, zeros, randn
Self CPU time total: 3.000us
cuda, zeros, zeros
Self CPU time total: 3.000us
cuda, randn, zt   
Self CUDA time total: 2.000us
cuda, zt,    randn
Self CUDA time total: 2.000us
error: dimension -1 does not have size 3

(1, 4) (2, 4, 8, 16, 32, 64, 3)
cuda, randn, randn
Self CPU time total: 4.000us
cuda, randn, zeros
Self CPU time total: 3.000us
cuda, zeros, randn
Self CPU time total: 3.000us
cuda, zeros, zeros
Self CPU time total: 3.000us
cuda, randn, zt   
Self CUDA time total: 2.000us
cuda, zt,    randn
Self CUDA time total: 2.000us
error: The size of tensor a (4) must match the size of tensor b (3) at non-singleton dimension 6

(4, 3) (1, 4)
cuda, randn, randn
Self CPU time total: 3.000us
cuda, randn, zeros
Self CPU time total: 3.000us
cuda, zeros, randn
Self CPU time total: 3.000us
cuda, zeros, zeros
Self CPU time total: 3.000us
cuda, randn, zt   
Self CUDA time total: 2.000us
cuda, zt,    randn
Self CUDA time total: 2.000us
error: The size of tensor a (3) must match the size of tensor b (4) at non-singleton dimension 1

(4, 3) (4, 3)
cuda, randn, randn
False
Self CUDA time total: 3.000us
cuda, randn, zeros
False
Self CUDA time total: 3.000us
cuda, zeros, randn
False
Self CUDA time total: 3.000us
cuda, zeros, zeros
False
Self CUDA time total: 3.000us
cuda, randn, zt   
True
Self CUDA time total: 2.000us
cuda, zt,    randn
True
Self CUDA time total: 2.000us

(4, 3) (4, 1)
cuda, randn, randn
False
Self CUDA time total: 3.000us
cuda, randn, zeros
False
Self CUDA time total: 2.000us
cuda, zeros, randn
False
Self CUDA time total: 2.000us
cuda, zeros, zeros
False
Self CUDA time total: 3.000us
cuda, randn, zt   
True
Self CUDA time total: 2.000us
cuda, zt,    randn
True
Self CUDA time total: 2.000us

(4, 3) (2, 4, 8, 16, 32, 64, 3)
cuda, randn, randn
Self CPU time total: 4.000us
cuda, randn, zeros
Self CPU time total: 4.000us
cuda, zeros, randn
Self CPU time total: 3.000us
cuda, zeros, zeros
Self CPU time total: 3.000us
cuda, randn, zt   
Self CUDA time total: 2.000us
cuda, zt,    randn
Self CUDA time total: 2.000us
error: The size of tensor a (4) must match the size of tensor b (64) at non-singleton dimension 5

(4, 1) (1, 4)
cuda, randn, randn
Self CPU time total: 3.000us
cuda, randn, zeros
Self CPU time total: 3.000us
cuda, zeros, randn
Self CPU time total: 3.000us
cuda, zeros, zeros
Self CPU time total: 3.000us
cuda, randn, zt   
Self CUDA time total: 2.000us
cuda, zt,    randn
Self CUDA time total: 2.000us
error: dimension -1 does not have size 3

(4, 1) (4, 3)
cuda, randn, randn
False
Self CUDA time total: 3.000us
cuda, randn, zeros
False
Self CUDA time total: 3.000us
cuda, zeros, randn
False
Self CUDA time total: 3.000us
cuda, zeros, zeros
False
Self CUDA time total: 3.000us
cuda, randn, zt   
True
Self CUDA time total: 2.000us
cuda, zt,    randn
True
Self CUDA time total: 2.000us

(4, 1) (4, 1)
cuda, randn, randn
Self CPU time total: 4.000us
cuda, randn, zeros
Self CPU time total: 3.000us
cuda, zeros, randn
Self CPU time total: 3.000us
cuda, zeros, zeros
Self CPU time total: 3.000us
cuda, randn, zt   
Self CUDA time total: 2.000us
cuda, zt,    randn
Self CUDA time total: 2.000us
error: dimension -1 does not have size 3

(4, 1) (2, 4, 8, 16, 32, 64, 3)
cuda, randn, randn
Self CPU time total: 3.000us
cuda, randn, zeros
Self CPU time total: 3.000us
cuda, zeros, randn
Self CPU time total: 3.000us
cuda, zeros, zeros
Self CPU time total: 2.000us
cuda, randn, zt   
Self CUDA time total: 2.000us
cuda, zt,    randn
Self CUDA time total: 2.000us
error: The size of tensor a (4) must match the size of tensor b (64) at non-singleton dimension 5

(2, 4, 8, 16, 32, 64, 3) (1, 4)
cuda, randn, randn
Self CPU time total: 3.000us
cuda, randn, zeros
Self CPU time total: 3.000us
cuda, zeros, randn
Self CPU time total: 3.000us
cuda, zeros, zeros
Self CPU time total: 3.000us
cuda, randn, zt   
Self CUDA time total: 2.000us
cuda, zt,    randn
Self CUDA time total: 2.000us
error: The size of tensor a (3) must match the size of tensor b (4) at non-singleton dimension 6

(2, 4, 8, 16, 32, 64, 3) (4, 3)
cuda, randn, randn
Self CPU time total: 3.000us
cuda, randn, zeros
Self CPU time total: 3.000us
cuda, zeros, randn
Self CPU time total: 3.000us
cuda, zeros, zeros
Self CPU time total: 3.000us
cuda, randn, zt   
Self CUDA time total: 2.000us
cuda, zt,    randn
Self CUDA time total: 2.000us
error: The size of tensor a (64) must match the size of tensor b (4) at non-singleton dimension 5

(2, 4, 8, 16, 32, 64, 3) (4, 1)
cuda, randn, randn
Self CPU time total: 3.000us
cuda, randn, zeros
Self CPU time total: 3.000us
cuda, zeros, randn
Self CPU time total: 3.000us
cuda, zeros, zeros
Self CPU time total: 3.000us
cuda, randn, zt   
Self CUDA time total: 2.000us
cuda, zt,    randn
Self CUDA time total: 2.000us
error: The size of tensor a (64) must match the size of tensor b (4) at non-singleton dimension 5

(2, 4, 8, 16, 32, 64, 3) (2, 4, 8, 16, 32, 64, 3)
cuda, randn, randn
False
Self CUDA time total: 252.000us
cuda, randn, zeros
False
Self CUDA time total: 251.000us
cuda, zeros, randn
False
Self CUDA time total: 253.000us
cuda, zeros, zeros
False
Self CUDA time total: 251.000us
cuda, randn, zt   
True
Self CUDA time total: 2.000us
cuda, zt,    randn
True
Self CUDA time total: 2.000us
```

</details>

Lezcano commented:
Could you please run the benchmarks with the benchmarking tool, rather than the profiler?

nkaretnikov commented:
yep, this seems more consistent. you can see that zt times are always constant regardless of input shape.

```py
import torch
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare

print('Using pytorch %s' % (torch.__version__))

shapes = [(128, 3), (256, 128, 3), (2, 512, 256, 3), (2, 64, 256, 128, 3)]
results = {}
repeats = 10
dtype = torch.double

with_zt = 'with_zt'
without_zt = 'without_zt'

inputs = (
    (lambda input: torch._efficientzerotensor(input.shape, dtype=input.dtype, device=input.device, requires_grad=True), with_zt),
    (lambda input: torch.zeros_like(input), without_zt),
)

for (input_fn, label) in inputs:
    results[label] = []
    for device in ['cpu', 'cuda']:
        for shape in shapes:
            input = torch.randn(*shape, dtype=dtype, device=device, requires_grad=True)
            input = input_fn(input)
            other = torch.randn(*shape, dtype=dtype, device=device, requires_grad=True)

            def fn(input, other):
                return torch.linalg.cross(input, other)

            tasks = [("fn(input, other)", label)]
            timers = [Timer(stmt=stmt, label=f"linalg_cross: zt input?: {input._is_zerotensor()}, device: {device}", sub_label=f"{(shape)}", description=desc, globals=globals()) for stmt, desc in tasks]

            max_len = 80
            for i, timer in enumerate(timers * repeats):
                results[label].append(
                    timer.blocked_autorange()
                )
                s = f"{i + 1} / {len(timers) * repeats}, {label}, {device}, {shape}"
                pad = max_len - len(s)
                print(f"\r{s}" + pad * " ", end="", flush=True)

print()

comparison = Compare(results[with_zt] + results[without_zt])
comparison.print()
```

output:

```
Using pytorch 1.12.0a0+git7e13bf8
[ linalg_cross: zt input?: True, device: cpu ]
                            |  with_zt
1 threads: ---------------------------
      (128, 3)              |    6.6  
      (256, 128, 3)         |    6.7  
      (2, 512, 256, 3)      |    6.9  
      (2, 64, 256, 128, 3)  |    7.0  

Times are in microseconds (us).

[ linalg_cross: zt input?: True, device: cuda ]
                            |  with_zt
1 threads: ---------------------------
      (128, 3)              |    13.0 
      (256, 128, 3)         |    13.1 
      (2, 512, 256, 3)      |    13.3 
      (2, 64, 256, 128, 3)  |    13.3 

Times are in microseconds (us).

[ linalg_cross: zt input?: False, device: cpu ]
                            |  without_zt
1 threads: ------------------------------
      (128, 3)              |        5.7 
      (256, 128, 3)         |      474.5 
      (2, 512, 256, 3)      |     8181.1 
      (2, 64, 256, 128, 3)  |   622691.1 

Times are in microseconds (us).

[ linalg_cross: zt input?: False, device: cuda ]
                            |  without_zt
1 threads: ------------------------------
      (128, 3)              |       8.7  
      (256, 128, 3)         |       8.8  
      (2, 512, 256, 3)      |      66.4  
      (2, 64, 256, 128, 3)  |    1014.5  

Times are in microseconds (us).
```

nkaretnikov commented:
@anjali411 @Lezcano updated benchmarks, ci is green. please merge if ok

anjali411 reviewed (approved):
LGTM thanks @nkaretnikov

anjali411 commented:
```
2022-05-10T02:45:25.2556906Z   test_variant_consistency_eager_eig_cuda_complex64 (__main__.TestCommonCUDA) ... python: /opt/conda/conda-bld/magma-cuda113_1619629459349/work/interface_cuda/interface.cpp:806: void magma_queue_create_internal(magma_device_t, magma_queue**, const char*, const char*, int): Assertion `queue->dAarray__ != __null' failed.
2022-05-10T02:45:27.5191170Z Traceback (most recent call last):
2022-05-10T02:45:27.5191531Z   File "test/run_test.py", line 1072, in <module>
2022-05-10T02:45:27.5195727Z     main()
2022-05-10T02:45:27.5196096Z   File "test/run_test.py", line 1050, in main
2022-05-10T02:45:27.5198139Z     raise RuntimeError(err_message)
2022-05-10T02:45:27.5198684Z RuntimeError: test_ops failed! Received signal: SIGIOT
2022-05-10T02:45:28.5700480Z + cleanup
```
I couldn't find this log in an upstream failure. could you verify if this failure is related to the PR?

Lezcano commented:
at first look that seems unrelated to this PR, as `linalg_cross` does not use magma internally. Any idea of what could be going on here @IvanYashchuk @xwang233 ?

anjali411 commented:
I see that this test is also failing on other PRs so seems unrelated

anjali411 commented:
@pytorchbot merge this please
